File: kotlin-null-writable/karakum/inheritanceModifiers/modifyMethodInheritance.js
import ts from "typescript"

export default (node) => {
    if (
        ts.isMethodDeclaration(node)
        && ts.isIdentifier(node.name)
        && (node.name.text === "_write"
            || node.name.text === "_writev"
        )
        && node.parent
        && node.parent.name?.text === "NullWritable"
    ) {
        return "override"
    }

    return null
}

File: kotlin-null-writable/karakum/src/jsMain/kotlin/nullwritable/karakum/inheritanceModifiers/modifyMethodInheritance.kt
package nullwritable.karakum.inheritanceModifiers

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.InheritanceModifierContext
import io.github.sgrishchenko.karakum.util.getParentOrNull
import typescript.Node
import typescript.isClassDeclaration
import typescript.isIdentifier
import typescript.isMethodDeclaration

fun modifyMethodInheritance(node: Node, context: InheritanceModifierContext) = nullable {
    ensure(isMethodDeclaration(node))

    val name = node.name
    ensure(isIdentifier(name))
    ensure(
        name.text == "_write"
                || name.text == "_writev"
    )

    val classNode = ensureNotNull(node.getParentOrNull())
    ensure(isClassDeclaration(classNode))
    ensure(classNode.name?.text == "NullWritable")

    "override"
}
